mllib/src/main/scala/org/apache/spark/ml/feature/VectorSlicer.scala

```diff
@@ -0,0 +1,153 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.attribute.AttributeGroup
+import org.apache.spark.ml.param.shared.{HasInputCol, HasOutputCol}
+import org.apache.spark.ml.param.{IntArrayParam, ParamMap, StringArrayParam}
+import org.apache.spark.ml.util.{Identifiable, SchemaUtils}
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.StructType
+
+/**
+ * :: Experimental ::
+ * Given either indices or names, it takes a vector column and output a vector column with the
+ * specified subset of features. Note that the vector column should contain ML [[Attribute]].
+ */
+@Experimental
+final class VectorSlicer(override val uid: String)
+  extends Transformer with HasInputCol with HasOutputCol {
+
+  def this() = this(Identifiable.randomUID("vectorSlicer"))
+
+  /**
+   * An array of indices to select features from a vector column.
+   * @group param
+   */
+  val selectedIndices = new IntArrayParam(this, "selectedIndices",
+    "An array of indices to select features from a vector column",
+    (x: Array[Int]) => if (x.isEmpty) true else x.min >= 0)
+
+  setDefault(selectedIndices -> Array.empty[Int])
+
+  /** @group getParam */
+  def getSelectedIndices: Array[Int] = getOrDefault(selectedIndices)
+
+  /** @group setParam */
+  def setSelectedIndices(value: Array[Int]): this.type = set(selectedIndices, value)
+
+  /**
+   * An array of feature names to select features from a vector column.
+   * @group param
+   */
+  val selectedNames = new StringArrayParam(this, "selectedNames",
+    "An array of feature names to select features from a vector column")
+
+  setDefault(selectedNames -> Array.empty[String])
+
+  /** @group getParam */
+  def getSelectedNames: Array[String] = getOrDefault(selectedNames)
+
+  /** @group setParam */
+  def setSelectedNames(value: Array[String]): this.type = set(selectedNames, value)
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Slice a dense vector with an array of indices.
+   */
+  private def selectColumns(indices: Array[Int], features: DenseVector): Vector = {
+    Vectors.dense(indices.map(features.apply))
+  }
+
+  /**
+   * Slice a sparse vector with a set of indices.
+   */
+  private def selectColumns(indices: Set[Int], features: SparseVector): Vector = {
+    Vectors.sparse(
+      indices.size, features.indices.zip(features.values).filter(x => indices.contains(x._1)))
```

Comment:
The resulting SparseVector will use the old indices, I think.  But it should use updated indices based on the "indices" argument.